service/src/DCGMIOGroup.cpp

```diff
@@ -0,0 +1,541 @@
+        }
+        else {
+    #ifdef GEOPM_DEBUG
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": Handling not defined for " +
+                            signal_name, GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+    #endif
+        }
+        return result;
+    }
+
+    // Write to the control immediately, bypassing write_batch()
+    void DCGMIOGroup::write_control(const std::string &control_name, int domain_type, int domain_idx, double setting)
+    {
+        if (!is_valid_control(control_name)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": " + control_name +
+                            " not valid for DCGMIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_type != control_domain_type(control_name)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": " + control_name + ": domain_type must be " +
+                            std::to_string(control_domain_type(control_name)),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_idx < 0 || domain_idx >= m_platform_topo.num_domain(GEOPM_DOMAIN_BOARD)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": domain_idx out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        if (control_name == "DCGM::FIELD_UPDATE_RATE") {
+            m_dcgm_device_pool.update_rate(setting*1e6);
+        }
+        else if (control_name == "DCGM::MAX_STORAGE_TIME") {
+            m_dcgm_device_pool.max_storage_time(setting);
+        }
+        else if (control_name == "DCGM::MAX_SAMPLES") {
+            m_dcgm_device_pool.max_samples(setting);
+        }
+        else {
+    #ifdef GEOPM_DEBUG
+                throw Exception("DCGMIOGroup::" + std::string(__func__) + "Handling not defined for "
+                                + control_name, GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+    #endif
+        }
+    }
+
+    // Implemented to allow an IOGroup to save platform settings before starting
+    // to adjust them
+    void DCGMIOGroup::save_control(void)
+    {
+    }
+
+    // Implemented to allow an IOGroup to restore previously saved
+    // platform settings
+    void DCGMIOGroup::restore_control(void)
+    {
+        m_dcgm_device_pool.update_rate(1*1e6); //1 second
+        m_dcgm_device_pool.max_samples(0); //0 = no limit
+        m_dcgm_device_pool.max_storage_time(10); //10 seconds.
+    }
+
+    void DCGMIOGroup::save_control(const std::string &save_path)
+    {
+        throw Exception("DCGMIOGroup::save_control()",
+                        GEOPM_ERROR_NOT_IMPLEMENTED, __FILE__, __LINE__);
+    }
+
+    void DCGMIOGroup::restore_control(const std::string &save_path)
+    {
+        throw Exception("DCGMIOGroup::restore_control()",
+                        GEOPM_ERROR_NOT_IMPLEMENTED, __FILE__, __LINE__);
+    }
```

Comment:
Since this IOGroup does not require any external state to save (noop) or restore (values compiled into binary), these should be implemented as a noop and a passthrough to the file free restore_control() method.